File: android/src/main/java/com/reveila/android/data/RoomRepository.java
package com.reveila.android.data;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.reveila.android.db.GenericDao;
import com.reveila.android.db.GenericEntity;
import com.reveila.data.Entity;
import com.reveila.data.EntityMapper;
import com.reveila.data.Filter;
import com.reveila.data.Page;
import com.reveila.data.Repository;
import com.reveila.data.Sort;

import java.util.ArrayList;
import java.util.Collection;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.UUID;
import java.util.stream.Collectors;

/**
 * Java implementation of the Room-backed repository.
 * Handles the mapping between generic Reveila Entities and Room GenericEntities.
 */
public class RoomRepository implements VaultRepository, Repository<Entity, Map<String, Map<String, Object>>> {

    private final String entityType;
    private final GenericDao dao;
    private final ObjectMapper objectMapper;
    
    // Equivalent to the Kotlin object : TypeReference...
    private final TypeReference<Map<String, Object>> mapTypeRef = new TypeReference<>() {};

    public RoomRepository(String entityType, GenericDao dao) {
        this.entityType = entityType;
        this.dao = dao;
        this.objectMapper = EntityMapper.getObjectMapper();
    }

    @Override
    public String getType() {
        return entityType;
    }

    @Override
    public Entity store(Entity entity) {
        try {
            String id = extractId(entity.getKey());
            String attributesJson = objectMapper.writeValueAsString(entity.getAttributes());
            GenericEntity genericEntity = new GenericEntity(id, entityType, attributesJson);
            dao.insert(genericEntity);
            return entity;
        } catch (Exception e) {
            throw new RuntimeException("Failed to store entity: " + entityType, e);
        }
    }

    @Override
    public Optional<Entity> fetchById(Map<String, Map<String, Object>> idMap) {
        String id = extractId(idMap);
        GenericEntity genericEntity = dao.findById(id, entityType);
        if (genericEntity != null) {
            return Optional.of(mapToEntity(genericEntity));
        }
        return Optional.empty();
    }

    @Override
    public void disposeById(Map<String, Map<String, Object>> idMap) {
        String id = extractId(idMap);
        dao.deleteById(id, entityType);
    }

    @Override
    public List<Entity> storeAll(Collection<Entity> entities) {
        List<GenericEntity> genericEntities = entities.stream().map(entity -> {
            try {
                String id = extractId(entity.getKey());
                String attributesJson = objectMapper.writeValueAsString(entity.getAttributes());
                return new GenericEntity(id, entityType, attributesJson);
            } catch (Exception e) {
                throw new RuntimeException("Serialization failed during batch store", e);
            }
        }).collect(Collectors.toList());

        dao.insertAll(genericEntities);
        return new ArrayList<>(entities);
    }

    @Override
    public List<Entity> fetchAll() {
        return dao.findByType(entityType).stream()
                .map(this::mapToEntity)
                .collect(Collectors.toList());
    }

    @Override
    public Page<Entity> fetchPage(Filter filter, Sort sort, List<String> fetches, int page, int size, boolean includeCount) {
        int offset = page * size;
        List<Entity> items = dao.fetchPage(entityType, size, offset).stream()
                .map(this::mapToEntity)
                .collect(Collectors.toList());

        Long count = includeCount ? dao.countByType(entityType) : null;
        boolean hasNext = items.size() == size;

        return new Page<>(items, page, size, hasNext, count);
    }

    @Override
    public long count() {
        return dao.countByType(entityType);
    }

    @Override
    public boolean hasId(Map<String, Map<String, Object>> idMap) {
        String id = extractId(idMap);
        return dao.exists(id, entityType) > 0;
    }

    @Override
    public void commit() {
        // Room auto-commits transactions on DAO methods.
        // For manual transaction control, you would use database.runInTransaction()
    }

    private String extractId(Map<String, Map<String, Object>> idMap) {
        if (idMap == null || idMap.isEmpty()) {
            return UUID.randomUUID().toString();
        }
        
        if (idMap.containsKey("id")) {
            Map<String, Object> idPart = idMap.get("id");
            if (idPart != null && idPart.containsKey("value")) {
                return String.valueOf(idPart.get("value"));
            }
        }
        
        // Fallback to first available value
        Map<String, Object> firstEntry = idMap.values().iterator().next();
        return String.valueOf(firstEntry.get("value"));
    }

    private Entity mapToEntity(GenericEntity genericEntity) {
        Map<String, Object> attributes;
        try {
            attributes = objectMapper.readValue(genericEntity.getAttributesJson(), mapTypeRef);
        } catch (Exception e) {
            attributes = new HashMap<>();
        }

        Map<String, Map<String, Object>> key = new HashMap<>();
        Map<String, Object> idPart = new HashMap<>();
        idPart.put("value", genericEntity.getId());
        key.put("id", idPart);

        return new Entity(entityType, key, attributes);
    }

    @Override
    public Map<String, Long> getIndexedFiles() {
        // Logic to return a map of URI -> LastModified from your DAO
        return new java.util.HashMap<>(); 
    }

    @Override
    public void insertFact(String subject, String type, String predicate, String object, String objectType, String metadata) {
        // Map this to your 'store' or a specific fact-insertion DAO method
    }

    @Override
    public void markFileAsIndexed(String uri, long timestamp) {
        // Update the 'last indexed' timestamp in your DB
    }

    @Override
    public void saveSecret(String data) {
        // Implementation for the Kill Switch or sensitive data storage
    }
}

File: android/src/main/kotlin/com/reveila/android/data/RoomRepository.kt
package com.reveila.android.data

import com.fasterxml.jackson.core.type.TypeReference
import com.reveila.android.db.GenericDao
import com.reveila.android.db.GenericEntity
import com.reveila.data.Entity
import com.reveila.data.EntityMapper
import com.reveila.data.Filter
import com.reveila.data.Page
import com.reveila.data.Repository
import com.reveila.data.Sort
import java.util.Optional
import java.util.UUID

class RoomRepository(
    private val entityType: String,
    private val dao: GenericDao
) : Repository<Entity, MutableMap<String, MutableMap<String, Any>>> {

    private val objectMapper = EntityMapper.getObjectmapper()
    private val mapTypeRef = object : TypeReference<MutableMap<String, Any>>() {}

    override fun getType(): String {
        return entityType
    }

    override fun store(entity: Entity): Entity {
        val id = extractId(entity.key)
        val attributesJson = objectMapper.writeValueAsString(entity.attributes)
        val genericEntity = GenericEntity(id, entityType, attributesJson)
        dao.insert(genericEntity)
        return entity
    }

    override fun fetchById(idMap: MutableMap<String, MutableMap<String, Any>>): Optional<Entity> {
        val id = extractId(idMap)
        val genericEntity = dao.findById(id, entityType)
        return if (genericEntity != null) {
            Optional.of(mapToEntity(genericEntity))
        } else {
            Optional.empty()
        }
    }

    override fun disposeById(idMap: MutableMap<String, MutableMap<String, Any>>) {
        val id = extractId(idMap)
        dao.deleteById(id, entityType)
    }

    override fun storeAll(entities: MutableCollection<Entity>): MutableList<Entity> {
        val genericEntities = entities.map { entity ->
            val id = extractId(entity.key)
            val attributesJson = objectMapper.writeValueAsString(entity.attributes)
            GenericEntity(id, entityType, attributesJson)
        }
        dao.insertAll(genericEntities)
        return entities.toMutableList()
    }

    override fun fetchAll(): MutableList<Entity> {
        return dao.findByType(entityType).map { mapToEntity(it) }.toMutableList()
    }

    override fun fetchPage(
        filter: Filter?,
        sort: Sort?,
        fetches: MutableList<String>?,
        page: Int,
        size: Int,
        includeCount: Boolean
    ): Page<Entity> {
        val offset = page * size
        val items = dao.fetchPage(entityType, size, offset).map { mapToEntity(it) }
        val count = if (includeCount) dao.countByType(entityType) else null
        val hasNext = items.size == size
        
        return Page(items, page, size, hasNext, count)
    }

    override fun count(): Long {
        return dao.countByType(entityType)
    }

    override fun hasId(idMap: MutableMap<String, MutableMap<String, Any>>): Boolean {
        val id = extractId(idMap)
        return dao.exists(id, entityType) > 0
    }

    override fun commit() {
        // Room auto-commits transactions on DAO methods
    }

    private fun extractId(idMap: MutableMap<String, MutableMap<String, Any>>?): String {
        if (idMap == null || idMap.isEmpty()) return UUID.randomUUID().toString()
        if (idMap.containsKey("id")) {
            return idMap["id"]?.get("value").toString()
        }
        return idMap.values.iterator().next()["value"].toString()
    }

    private fun mapToEntity(genericEntity: GenericEntity): Entity {
        val attributes: MutableMap<String, Any> = try {
            objectMapper.readValue(genericEntity.attributesJson, mapTypeRef)
        } catch (e: Exception) {
            mutableMapOf()
        }
        
        // Ensure id is part of attributes or key
        val key = mutableMapOf<String, MutableMap<String, Any>>()
        val idPart = mutableMapOf<String, Any>("value" to genericEntity.id)
        key["id"] = idPart
        
        return Entity(entityType, key, attributes)
    }
}
